web/web.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2017 Caicloud Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package web
+
+import (
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/caicloud/nirvana/definition"
+	"github.com/caicloud/nirvana/log"
+	"github.com/caicloud/nirvana/router"
+)
+
+// Server handles HTTP requests.
+//
+// Workflow:
+//            Server.ServeHTTP()
+//          ----------------------
+//          ↓                    ↑
+// |-----Filters------|          ↑
+//          ↓                    ↑
+// |---Router Match---|          ↑
+//          ↓                    ↑
+// |-------------Middlewares------------|
+//          ↓                    ↑
+// |-------------Executor---------------|
+//          ↓                    ↑
+// |-ParameterGenerators-|-TypeHandlers-|
+//          ↓                    ↑
+// |------------User Function-----------|
+type Server interface {
+	// AddDescriptors adds descriptors to router.
+	AddDescriptors(descriptors ...definition.Descriptor) error
+	// AddFilters add filters to filter requests.
+	AddFilters(filters ...Filter)
+	// SetLogger sets logger to server.
+	SetLogger(logger log.Logger)
+	// SetModifier sets definition modifier.
+	SetModifier(m DefinitionModifier)
+	// ServeHTTP is used to handle request.
+	ServeHTTP(resp http.ResponseWriter, req *http.Request)
+}
+
+// NewDefaultServer creates a server with default config.
+// It uses three filters: RedirectTrailingSlash, ParseRequestForm, FillLeadingSlash.
+// It uses one modifier: FirstContextParameter.
+func NewDefaultServer() Server {
+	s := NewServer()
+	s.AddFilters(RedirectTrailingSlash(), ParseRequestForm(), FillLeadingSlash())
+	s.SetModifier(FirstContextParameter())
```

Comment:
If `FirstContextParameter()` is not set, server will panic. So `FirstContextParameter()` modifier should always be added.
The problem is `NewServer`  does not set this modifier. Please fix it.

Reply:
I'll add more comments to function `NewServer()`. The server is not used for common scene. It provides a complete capacity to customize a specific server.